### PR TITLE
Feat: support post message

### PR DESCRIPTION
**Pull request by fabian4**
https://github.com/wechaty/wechaty/issues/2245

### CODE REVIEW

**Comment by huan** [comment]
Which issue does this PR related to?

Please always link them together.

**Comment by fabian4** [comment]
ping @huan @su-chang

File: src/help/message.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ import PuppetWalnut from '../puppet-walnut.js'
 import type { FileBoxInterface } from 'file-box'
 import type { MessageItem } from './struct.js'
 import type * as PUPPET from 'wechaty-puppet'
+import type { SayablePayload } from 'wechaty-puppet/dist/cjs/src/schemas/sayable'
```
Comment (huan) [inline src/help/message.ts:9]: Never use deep import.

Use `PUPPET.payloads.Sayable` instead.

File: src/help/message.ts [hunk 2]
```diff
@@ -33,6 +34,44 @@ export async function sendFileMessage (to: string, file: FileBoxInterface) {
   })
 }
 
+export async function sendPostMessage (to: string, postPayload: PUPPET.payloads.Post) {
```
Comment (huan) [inline src/help/message.ts:37]: Rename `to` to `contactId` to make the param name more clear.

File: src/help/message.ts [hunk 3]
```diff
@@ -33,6 +34,44 @@ export async function sendFileMessage (to: string, file: FileBoxInterface) {
...
+  const title = postPayload.sayableList[0] as SayablePayload
+  const description = postPayload.sayableList[1] as SayablePayload
+  const img = postPayload.sayableList[2] as SayablePayload
```
Comment (huan) [inline src/help/message.ts:40]: If you need a `title + description + image` combo, then it seems that you can use a UrlLink Payload.

Your payload can be

```ts
PUPPET.payloads.sayable.url({
  url: 'https://google.com',
  title: 'Google',
  description: 'description',
  thumbnailUrl: 'https://xxx.com/img.jpg',
})
```

The only problem is that the `url` and `title` in the payload are required, which means you need to set those two values. (a '' might be a workaround, however, more suggestions are welcome.)

File: src/help/message.ts [hunk 4]
```diff
@@ -33,6 +34,44 @@ export async function sendFileMessage (to: string, file: FileBoxInterface) {
...
+  if (title.type !== 'Text' || description.type !== 'Text' || img.type !== 'Attachment') {
+    throw new Error('Wrong Post!!!')
+  }
+
+  const fileItem = await uploadFile(true, (<FileBoxInterface>img.payload.filebox))
+
+  sendMessage(to, {
```
Comment (su-chang) [inline src/help/message.ts:47]: The params is complicated indeed, it's better to add a new function to generate it.

BTW, test case will make the function more reliable.

File: src/help/message.ts [hunk 5]
```diff
@@ -33,6 +34,44 @@ export async function sendFileMessage (to: string, file: FileBoxInterface) {
   })
 }
 
+export async function sendPostMessage (to: string, postPayload: PUPPET.payloads.Post) {
+  const title = postPayload.sayableList[0] as SayablePayload
+  const description = postPayload.sayableList[1] as SayablePayload
+  const img = postPayload.sayableList[2] as SayablePayload
+  if (title.type !== 'Text' || description.type !== 'Text' || img.type !== 'Attachment') {
+    throw new Error('Wrong Post!!!')
```
Comment (su-chang) [inline src/help/message.ts:42]: Error message without detail info, a little help for debugging only.

File: src/puppet-walnut.ts [hunk 6]
```diff
@@ -211,7 +210,7 @@ class PuppetWalnut extends PUPPET.Puppet {
       case MessageRawType.other:
         res.type = PUPPET.types.Message.Attachment
         res.text = 'file'
-        if (file.contentType === 'text/vcard') {
+        if (files[0]!.contentType === 'text/vcard') {
```
Comment (su-chang) [inline src/puppet-walnut.ts:213]: Please think more about using `!`.

File: src/help/message.ts [hunk 7]
```diff
@@ -33,6 +34,44 @@ export async function sendFileMessage (to: string, file: FileBoxInterface) {
   })
 }
 
+export async function sendPostMessage (to: string, postPayload: PUPPET.payloads.Post) {
+  const title = postPayload.sayableList[0] as SayablePayload
```
Comment (su-chang) [inline src/help/message.ts:38]: It seems that the attrs of `postPayload.sayableList` must in order?

**Review by su-chang — CHANGES_REQUESTED** [review]
BTW, all commits with the same commit message is not a good habit.

**Review by su-chang — APPROVED** [review]
LGTM

File: src/help/message.ts [hunk 8]
```diff
@@ -7,37 +7,75 @@ import type { FileBoxInterface } from 'file-box'
 import type { MessageItem } from './struct.js'
 import type * as PUPPET from 'wechaty-puppet'
 
-export function sendTextMessage (to: string, msg: string) {
-  sendMessage(to, {
+export function sendTextMessage (contactId: string, msg: string) {
+  sendMessage(contactId, {
```
Comment (huan) [inline src/help/message.ts:11]: Does this nested function necessary?

Embed the exact code to reduce the complexity if possible.

File: src/help/message.ts [hunk 9]
```diff
@@ -7,37 +7,75 @@ import type { FileBoxInterface } from 'file-box'
...
     contentText: 'geo:50.7311865,7.0914591;crs=gcj02;u=10;rcs-l=Qingfeng%20Steamed%20Dumpling%20Shop %20%F0%9F%8D%9A',
     contentType: contentType.text,
   })
 }
 
-export async function sendFileMessage (to: string, file: FileBoxInterface) {
+export async function sendFileMessage (contactId: string, file: FileBoxInterface) {
   const fileItem = await uploadFile(true, file)
-  sendMessage(to, {
+  sendMessage(contactId, {
```
Comment (huan) [inline src/help/message.ts:29]: Does this nested function necessary?

Embed the exact code to reduce the complexity if possible.

**Review by huan — APPROVED** [review]
I have no more suggestions.

Looking forward to seeing the Post support for walnut online soon.

Thank you very much for the implementations!

**Pull request merged**